<div style="border:solid green 2px; padding: 20px"> <h1 style="color:green; margin-bottom:20px">Reviewers comment v1</h1>

Hello Kelvin!

Im happy to review your project today 🙌

You can find my comments under the heading **«Review»**. I will categorize my comments in green, blue or red boxes like this:

<div class="alert alert-success">
    <b>Success:</b> if everything is done successfully
</div>
<div class="alert alert-warning">
    <b>Remarks:</b> if I can give some recommendations or ways to improve the project
</div>
<div class="alert alert-danger">
    <b>Needs fixing:</b> if the block requires some corrections. Work cant be accepted with the red comments
</div>

Please dont remove my comments :) If you have any questions dont hesitate to respond to my comments in a different section. 
<div class="alert alert-info"> <b>Student comments:</b> For example like this</div>    



<div style="border:solid green 2px; padding: 20px">
<b>Reviewer's comment v1:</b>
    
<b>Overall Feedback</b> 
    

Hello Kelvin,

You’ve submitted another project—great work! Your commitment to pushing through the challenges of this program is admirable.

After reviewing your submission, I’ve returned it with some feedback to help you make the necessary improvements.

You’ll find specific feedback in the notebook in the ‘Comments’ (`Reviewer's comment v1`) sections.

**What Was Great:**

- Data Loading and Preparation: You successfully loaded the dataset and correctly separated the features and target variable. This is a crucial first step, and you handled it appropriately.
- Model Training and Evaluation: Using a loop to train and evaluate each model is an efficient approach. You correctly used the accuracy_score metric to assess the model.
    
**Points to improve:**
- Could you please print the sizes of datasets after the split. 
    
Keep in mind that revisions are a normal and valuable part of the learning process. Use this feedback to refine your work and resubmit when you’re ready. I know you’re capable of great things, and I’m here to support you every step of the way. Keep going—you’re doing a great job! 🏄

And of course, if you have any questions along the way, remember that you can always reach out to your tutor for clarification.

<div style="border:solid green 2px; padding: 20px">
<b>Reviewer's comment v2:</b>
    
Thank you for going the extra mile in incorporating additional changes and improvements.

Wish you cool projects in the next sprints! ☘️ 

Additionally you could check some nice examples of ML notebooks on Kaggle: https://www.kaggle.com/learn/intro-to-machine-learning

# Megaline Mobile

Megaline is a mobile carrier looking to transition customers from outdated legacy plans to one of its newer offerings: the Smart or Ultra plan. To support this goal, I will be developing a classification model that can analyze user behavior and accurately.
Using a dataset containing behavioral data from current Smart and Ultra users, I will build and evaluate several machine learning models to predict which plan a customer is most likely to use. The objective was to achieve the highest possible accuracy, with a minimum performance threshold of 0.75 measured on the test dataset. I will proceed directly to model training, validation, tuning, and final evaluation to select the most effective solution for Megaline’s recommendation system.

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Great that you've added a short information about the project goal. It gives an overview of what you are going to achieve in this project.


In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LogisticRegression

In [2]:
df_mobile= pd.read_csv('/datasets/users_behavior.csv')

In [3]:
df_mobile.head()

,calls,minutes,messages,mb_used,is_ultra
0,40.0,311.90,83.0,19915.42,0
1,85.0,516.75,56.0,22696.96,0
2,77.0,467.66,86.0,21060.45,0
3,106.0,745.53,81.0,8437.39,1
4,66.0,418.74,1.0,14502.75,0


In [4]:
df_mobile.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [5]:
df_mobile.isnull().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Well done! Data have been successfully loaded and inspected.

# Split the source data into a training set, a validation set, and a test set

In [6]:
df_train, temp = train_test_split(df_mobile, test_size=0.4, random_state=42)
df_valid, df_test = train_test_split(temp, test_size=0.50, random_state=42)

features = ['calls', 'minutes', 'messages', 'mb_used']

x_train = df_train[features]
y_train = df_train['is_ultra']
x_valid = df_valid[features]
y_valid = df_valid['is_ultra']
x_test = df_test[features]
y_test = df_test['is_ultra']

print(f'x_train shape: {x_train.shape}, y_train shape: {y_train.shape}')
print(f'x_valid shape: {x_valid.shape}, y_valid shape: {y_valid.shape}')
print(f'x_test shape: {x_test.shape}, y_test shape: {y_test.shape}')

x_train shape: (1928, 4), y_train shape: (1928,)
x_valid shape: (643, 4), y_valid shape: (643,)
x_test shape: (643, 4), y_test shape: (643,)


<div class="alert alert-info"> <b>Student comments:printed the size of dataframes using shape and f-string   


<div class="alert alert-success">
<b>Reviewer's comment v2:</b>
    
Great everything is correct here! 


<div class="alert alert-danger">
<b>Reviewer's comment v1:</b>

Could you please print the sizes of dataframes using `shape` and `f-strings`

`print(f'The training set shape is {features_tr_val.shape} for features...')`</div>

<div class="alert alert-warning">
<b>Reviewer's comment v1:</b>
    
Consider using stratified splits (`stratify` parameter) to ensure that class distributions are similar across training, validation, and test sets.



 I split the original dataset into three parts: a training set, a validation set, and a test set. I first separated 40% of the data into a temporary set, then split that in half to create two equal-sized validation and test sets. This allowed me to train the models on one portion of the data, fine-tune and evaluate them using the validation set, and finally assess the chosen model’s performance on the test set. I also defined the feature columns and target variable (is_ultra) for each subset.

# Investigate the quality of different models by changing hyperparameters

# Decision Tree Classifier

In [7]:
dt_model = DecisionTreeClassifier(max_depth=5, random_state=42)
dt_model.fit(x_train, y_train)
y_pred_valid = dt_model.predict(x_valid)
valid_accuracy = accuracy_score(y_valid, y_pred_valid)
print(f"Validation Accuracy:", valid_accuracy)

Validation Accuracy: 0.7729393468118196


In [8]:
for depth in range(1, 11):
    tree = DecisionTreeClassifier(max_depth=depth, random_state=42)
    tree.fit(x_train, y_train)
    val_preds = tree.predict(x_valid)
    acc = accuracy_score(y_valid, val_preds)
    print(f"max_depth={depth} | Accuracy: {acc:.3f}")

max_depth=1 | Accuracy: 0.731
max_depth=2 | Accuracy: 0.782
max_depth=3 | Accuracy: 0.792
max_depth=4 | Accuracy: 0.781
max_depth=5 | Accuracy: 0.773
max_depth=6 | Accuracy: 0.776
max_depth=7 | Accuracy: 0.781
max_depth=8 | Accuracy: 0.796
max_depth=9 | Accuracy: 0.781
max_depth=10 | Accuracy: 0.795


In [9]:
# final_dt_model = DecisionTreeClassifier(max_depth=5, random_state=42)
# final_dt_model.fit(x_train, y_train)
# test_preds = final_dt_model.predict(x_test)
# test_accuracy = accuracy_score(y_test, test_preds)

# print(f"Test Accuracy: {test_accuracy:.3f}")

In this part of the project, I explored how changing the max_depth hyperparameter affects the performance of a Decision Tree Classifier. I started by training a model with max_depth=5, which gave me a validation accuracy of around 0.773. To find the best depth, I looped through values from 1 to 10 and trained a new model for each one. I then evaluated each model using the validation set and printed out the accuracy scores. From the results, I observed that the accuracy generally improved as the depth increased, with the best performance at max_depth=8, achieving an accuracy of 0.796. This tuning process helped me choose a depth that avoids underfitting or overfitting.

# Random Forest Classifier

In [10]:
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(x_train, y_train)
y_pred = rf_model.predict(x_valid)
accuracy = accuracy_score(y_valid, y_pred)

print(f"Validation Accuracy: {accuracy:.3f}")

Validation Accuracy: 0.802


In [11]:
for n in range(10, 101, 10):
    rf_model = RandomForestClassifier(n_estimators=n, random_state=42)
    rf_model.fit(x_train, y_train)
    y_pred = rf_model.predict(x_valid)
    accuracy = accuracy_score(y_valid, y_pred)
    
    print(f"n_estimators={n} | Accuracy: {accuracy:.3f}")

n_estimators=10 | Accuracy: 0.787
n_estimators=20 | Accuracy: 0.792
n_estimators=30 | Accuracy: 0.793
n_estimators=40 | Accuracy: 0.796
n_estimators=50 | Accuracy: 0.796
n_estimators=60 | Accuracy: 0.796
n_estimators=70 | Accuracy: 0.798
n_estimators=80 | Accuracy: 0.801
n_estimators=90 | Accuracy: 0.802
n_estimators=100 | Accuracy: 0.802


In [12]:
# final_model = RandomForestClassifier(n_estimators=100, random_state=42)
# final_model.fit(x_train, y_train)
# y_test_pred = final_model.predict(x_test)
# test_accuracy = accuracy_score(y_test, y_test_pred)

# print(f"Test Accuracy: {test_accuracy:.3f}")

 I evaluated the performance of a Random Forest Classifier by tuning the n_estimators hyperparameter, which controls the number of trees in the forest. I first trained a model with n_estimators=100 and got a validation accuracy of 0.802. I ran a loop from 10 to 100 in steps of 10, training a new model each time. I found that the accuracy gradually improved and started to stabilize around n_estimators=70, with the best results at n_estimators=90 and 100, both reaching an accuracy of 0.802. This experiment helped me confirm that using 100 estimators gives me strong and stable performance without significant overfitting or underfitting.

# Logistic Regression 

In [13]:
lr_model = LogisticRegression(max_iter=200, random_state=42)
lr_model.fit(x_train, y_train)
y_pred = lr_model.predict(x_valid)
accuracy = accuracy_score(y_valid, y_pred)

print(f"Validation Accuracy: {accuracy:.3f}")

Validation Accuracy: 0.740


In [14]:
# final_lr_model = LogisticRegression(max_iter=200, random_state=42)
# final_lr_model.fit(x_train, y_train)
# y_test_pred = final_model.predict(x_test)
# test_accuracy = accuracy_score(y_test, y_test_pred)

# print(f"Test Accuracy: {test_accuracy:.3f}")

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Great that you've used the validation set for hyperparameter tuning and model selection and only evaluate the final model on the test set.
    
In machine learning, overfitting occurs when a model learns the training data too well, including its noise and random fluctuations, hindering its ability to generalize to new data. Hyperparameters and model selection significantly impact a model's performance, but tuning them using the test set can lead to data leakage and overfitting. To avoid this, a validation set is used as a surrogate for the test set during development, providing an unbiased performance estimate.
    

 I trained a Logistic Regression model with max_iter=200 to make sure the model had enough iterations to converge. After fitting it on the training data, I evaluated its performance using the validation set. The model achieved a validation accuracy of 0.740, which was lower than both the Decision Tree and Random Forest models.

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Everything is correct here! Great that you've managed to check multiple models. 

Some possible minor improvements: 

- If you find yourself repeating similar blocks of code, consider writing a function. This will make your code more organized and easier to maintain.
- Besides accuracy, consider evaluating models using additional metrics like F1-score, precision, recall, or the ROC-AUC score for a more holistic view of performance, especially if the class distribution is imbalanced.
- Consider using GridSearchCV or RandomizedSearchCV from sklearn.model_selection for a more systematic hyperparameter search.

# Test the chosen model

I tested the chosen model—the Random Forest Classifier with 100 estimators—to evaluate how well it performs on completely unseen data. After tuning and comparing different models using the validation set, the Random Forest consistently gave the highest validation accuracy (0.802), and when tested on the actual test set, it achieved a strong test accuracy of 0.812.

In [15]:
final_model = RandomForestClassifier(n_estimators=100, random_state=42)
final_model.fit(x_train, y_train)
y_test_pred = final_model.predict(x_test)
test_accuracy = accuracy_score(y_test, y_test_pred)

print(f"Test Accuracy: {test_accuracy:.3f}")

Test Accuracy: 0.812


<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Great! Well above the required threshold. 

The final test was important because it helps me confirm that the model generalizes well beyond the training and validation sets. I chose Random Forest because it outperformed both the Decision Tree and Logistic Regression models in terms of accuracy, and it's also known for being robust to overfitting while handling complex patterns in the data. That makes it the best model to use for this problem.

<div class="alert alert-warning">
<b>Reviewer's comment v1:</b>
    
**Additional task: sanity check the model.** 
 
The use of `DummyClassifier` is a good practice for baseline performance and finalizing sanity check. You could read about it [here](https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html). In our case, we have a very skewed distribution of users across the two plans. As you can see, only 30% are enrolled in the Ultra plan, therefore it could be usefull to use a particular strategy (e.g., `most_frequent`) and comparing it against the chosen model’s performance.
    
```
from sklearn.dummy import DummyClassifier
# Initialize the DummyClassifier to predict the most frequent class
dummy_clf = DummyClassifier(strategy="most_frequent", random_state=0)
# Fit the dummy classifier on the training data
...
```

In [16]:
from sklearn.dummy import DummyClassifier

dummy_clf = DummyClassifier(strategy="most_frequent", random_state=0)
dummy_clf.fit(x_train, y_train)
dummy_pred = dummy_clf.predict(x_test)
dummy_accuracy = accuracy_score(y_test, dummy_pred)
print(f"Dummy Classifier Accuracy (most frequent): {dummy_accuracy:.3f}")

Dummy Classifier Accuracy (most frequent): 0.697


<div class="alert alert-info"> <b>Student comments: Used the DummClassifier the results came out for most frequent is 0.697 clearly shows that the Random Forest model performs significantly better.

<div class="alert alert-success">
<b>Reviewer's comment v2:</b>
    
👏 Fully agree. 


# General Conclusion

In this project, I analyzed user behavior data from Megaline Mobile to predict whether a customer would use the "Ultra" plan or not. I began by loading and exploring the dataset, checking for missing values and understanding the distribution of features such as calls, minutes, messages, and internet data usage.I split the data into training, validation, and test sets to ensure proper model evaluation. Then, I trained and compared three different classification models: Decision Tree, Random Forest, and Logistic Regression. For each model, I tuned relevant hyperparameters and evaluated their performance using the validation set.
The Random Forest Classifier provided the best performance with a validation accuracy of 0.802 and a final test accuracy of 0.812. It outperformed the Decision Tree and Logistic Regression models in terms of both consistency and accuracy. I selected the Random Forest model as the best option for this classification task because of its strong generalization ability, robustness to overfitting, and reliable performance on unseen data.



<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    

Great job on your overall conclusions and recommendations!  Your recommendations are well-thought and could be very valuable to the business.